libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,42 +112,121 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
             }
-            await base.ProcessHttpRequestAsync(request, cancellationToken);
+            await base.ProcessHttpRequestAsync(request, cancellationToken).ConfigureAwait(false);
         }
 
         public async Task<string> GetTokenAsync(bool forceRefresh = false)
         {
-            if (forceRefresh == false)
+            Task<OAuthResponse> oAuthTokenTask = null;
+            OAuthResponse oAuthToken = null;
+
+            // get tokenTask from cache 
+            lock (tokenTaskCache)
             {
-                // check the global cache for the token. If we have it, and it's valid, we're done. 
-                lock (_cacheSync)
+                // if we are being forced or don't have a token in our cache at all
+                if (forceRefresh || !tokenCache.TryGetValue(TokenCacheKey, out oAuthToken))
+                {
+                    // we will await this task, because we don't have a token and we need it
+                    oAuthTokenTask = _getCurrentTokenTask(forceRefresh: forceRefresh);
+                }
+                else
                 {
-                    bool foundToken = _cache.TryGetValue(TokenCacheKey, out OAuthResponse oAuthToken);
-                    if (foundToken)
+                    // we have an oAuthToken
+                    // check to see if our token is expired 
+                    if (IsTokenExpired(oAuthToken))
                     {
-                        // we have the token. Is it valid? 
-                        if (oAuthToken.expiration_time > DateTime.UtcNow)
+                        // it is, we should await the current task (someone could have already asked for a new token)
+                        oAuthTokenTask = _getCurrentTokenTask(forceRefresh: false);
+
+                        // if the task is completed and is the expired token, then we need to force a new one 
+                        // (This happens if bot has been 100% idle past the expiration point)
+                        if (oAuthTokenTask.Status == TaskStatus.RanToCompletion && oAuthTokenTask.Result.access_token == oAuthToken.access_token)
                         {
-                            return oAuthToken.access_token;
+                            oAuthTokenTask = _getCurrentTokenTask(forceRefresh: true);
                         }
                     }
+
+                    // always check the autorefresh
+                    CheckAutoRefreshToken();
+                }
+            }
+
+            // if we have an oAuthTokenTask then we need to await it
+            if (oAuthTokenTask != null)
+            {
+                oAuthToken = await oAuthTokenTask;
+                tokenCache[TokenCacheKey] = oAuthToken;
+            }
+
+            return oAuthToken?.access_token;
+        }
+
+        private void CheckAutoRefreshToken()
+        {
+            // get the current autorefreshTime for this key
+            DateTime refreshTime;
```

Comment:
You can inline declare within the `TryGetValue` so it's not just floating around out here:

```CSharp
if(autoRefreshTimes.TryGetValue(TokenCacheKey, out var refreshTime))
```

Reply:
BTW: The point of this feature is to reduce typing.  I don't think it helps with readibility and it clearly doesn't impact performance or runtime.

---
In reply to: [198880562](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198880562) [](ancestors = 198880562)